### PR TITLE
ft/1747-dataset-child-resources

### LINKED ISSUE #1747: FE - Ištekliaus tabas ištekliaus atvaizdavimo puslapyje
**Specifikacijos punktas:**
- https://github.com/atviriduomenys/katalogas/issues/1793
#

Užduoties tikslas ištekliaus atvaizdavimo puslapyje sukurti naują tab'ą:
Ištekliai - kuriame atvaizduojami vaikiniai ištekliai (vaikinių išteklių grandinėlė, ilgis iki 10 elementų)
Tam, kad būtų galima atvaizduoti, visus vaikinius išteklius (rekursinė grandinėlė), gali tekti modifikuoti ryšio "hasPart" sudarymą.

Vaikinių išteklių atranka:

- Atrenkami tėvinio išteklio ryšiai kurių ryšio tipas "hasPart" (subject - tėvinis, object - vaikinis)
- Atrenkami atrintkų ryšių ištekliai - šie ištekliai tiesioginiai vaikiniai ištekliai
- ir tokią atranką kartojame rekursiškai kol atrenkame visas vaikų grandinėles

-Tab'ą įterpti tarp Informacija ir Struktūra Image

Pastaba analitkui:
- Tą patį ryšį galima realizuoti pagrindinių ryšiu tarp tėvo ir vaiko su hasPart ir tarp vaiko ir tėvo su isPart, nepaisant to, kad šie du ryšiai logiškai aprašytų tokį patį ryšį, laikome kad hasPart yra pagrindinis ryšys ir isPart yra atvirkščias ryšys ir isPart ryšio nenaudojame, vaikų atrankai.

-

pastaba programuotojui:
- Atliekant užduotį, tab'o turinys gali būti perpanaudotas Išteklių atvaizdavimo puslapis arba Mano organizacijos Ištekliai, naujo tab'o turinio kurti nereikia, reikia užnaudoti esamą su kitu filtru.
- Uzklausa ryšiam atrinkti, ryšiai gali būti įvairūs, šio ticket scope domina tik 'hasPart'
- realizuoti grandinėle galima naudojant Materialized path tree
```

select
  dst.id, --subject
  dtt.title,
  rel.name,
  rel.uri,
  dstc.id, --object
  dstct.title
from
  dataset dst,
  dataset_translation dtt,
  dataset_relation drl,
  relation rel,
  dataset dstc,
  dataset_translation dstct
where
  -- access predicates
  drl.dataset_id = dst.id and
  dtt.master_id = dst.id and
  rel.id = drl.relation_id and
  dstc.id = drl.part_of_id and
  dstct.master_id = dstc.id and
  -- filter predicates
  dtt.language_code = 'lt' and
  dstct.language_code = 'lt' and
  rel.name  in ('part-of', 'hasPart')

```

# Priėmimo testavimo scenarijus 

### Testavimo eiga 

----
1. Atidaromas ištekliaus `test` atvaizdavimo puslapis
---

### Testuojama kaip 

Partnerio atstovas 

#### Sąlygos 

1. Egzistuoja informacinės sistemos tipo duomenų išteklius `test` ir jis priklauso Partnerio atstovo tvarkomai organizacijai

2. Partnerio atstovas turi teisę peržiūrėti duomenų ištekliaus duomenis

3. Ištekliui `test` egzistuoja visi jam įmanomi vaikiniai ištekliai `test_{m}` ir visi jų įmanomi vaikiniai ištekliai (rekursiškai)

#### Veiksmai 

1. Atidaromas ištekliaus `test` atvaizdavimo puslapis

2. Pasirenkama kortelė `Ištekliai`

#### Įvedami duomenys 

NA

#### Rezultatas

2. Matomi visi ištekliaus `test` vaikiniai ištekliai `test_{m}`

### PR DESCRIPTION
**Pull request by varzgalys**
# Child dataset resources
-----

## Abstract Resource model
Introduce Resource abstract model as concept of understanding among tech people and requirements.

## New relations (materialized path tree)
Introduce new dataset relation structure based on MP_Node side by already existing DatasetRelations, however they are tracking different relations and are NOT THE SAME THING. This relation will store from new scratch (not migrating any old relations)

Programatic relation management is implemented by [treebeard library  ](https://django-treebeard.readthedocs.io/en/latest/mp_tree.html) - each Dataset istance will `get_parent(), get_ancestors(), get_childs(), get_decendants()` and other methods (check docs).

### Dataset child resources list view (reused Dataset list view)
image

Together with new buttons create child resource under the select resource
image

Or selection inside of a Dataset form in case creating from root (my organization datasets)
image

## Admin to manage DatasetRelations (old ones)
image

## Refactor dataset URLs
Fix url design from `datasets/<org-id>/*` to `orgs/<org-id>/datasets/*`

### CODE REVIEW

File: vitrina/datasets/migrations/0015_sync_dataset_relations.py
Comment (Struchid) [inline vitrina/datasets/migrations/0015_sync_dataset_relations.py:1]: Can the migrations be combined?

File: vitrina/datasets/views.py [hunk 1]
```diff
@@ -3895,3 +3898,57 @@ def get_context_data(self, **kwargs):
...
+    def parent_dataset(self) -> Dataset:
+        return get_object_or_404(Dataset, pk=self.parent_dataset_id)
+
+    @cached_property
+    def organization(self) -> Organization:
+        return self.parent_dataset.organization
+
+    @property
+    def organization_id(self) -> int:
+        return self.parent_dataset.organization_id
```
Comment (Struchid) [inline vitrina/datasets/views.py]: Empty line at eof is missing

File: vitrina/locale/en/LC_MESSAGES/django.po [hunk 2]
```diff
@@ -583,6 +583,11 @@ msgstr ""
 "Is the institution you are representing responsible for opening the data to "
 "the public?"
 
+#, fuzzy
+#| msgid "Duomenų išteklius"
+msgid "Tėvynis išteklius"
+msgstr "Parent resource"
```
Comment (Struchid) [inline vitrina/locale/en/LC_MESSAGES/django.po]: Remove the fuzzy part:

```suggestion
msgid "Tėvynis išteklius"
msgstr "Parent resource"
```

File: vitrina/locale/en/LC_MESSAGES/django.po [hunk 3]
```diff
@@ -1405,6 +1418,11 @@ msgstr "Most relevant"
 msgid "Nepavyko nuskaityti failo"
 msgstr "Failed to read the file"
 
+#, fuzzy
+#| msgid "Pridėti duomenų išteklių"
+msgid "Pridėti vaikinių duomenų išteklių"
+msgstr "Add a resource"
```
Comment (Struchid) [inline vitrina/locale/en/LC_MESSAGES/django.po]: Translation is not correct + the fuzzy should be removed.

File: vitrina/locale/en/LC_MESSAGES/django.po [hunk 4]
```diff
@@ -5085,3 +5106,6 @@ msgid ""
 "Prisijunkite su šiuo elektroniniu paštu ir bandykite viisp autentifikaciją "
 "iš naujo"
 msgstr "Log in with this e-mail address and try viisp authentication again"
+
+#~ msgid "Sinchronizuoti medį"
+#~ msgstr "Synchronize tree"
```
Comment (Struchid) [inline vitrina/locale/en/LC_MESSAGES/django.po]: This is not needed

File: tests/datasets/test_views.py [hunk 5]
```diff
@@ -2095,6 +2094,31 @@ def test_dataset_create_public(app: DjangoTestApp):
     assert added_dataset.first().published is not None
     assert added_dataset.first().access_rights == Dataset.PUBLIC
 
+@pytest.mark.django_db
+def test_child_dataset_create_public(app: DjangoTestApp):
```
Comment (Struchid) [inline tests/datasets/test_views.py:2098]: Only one test case for these changes? No sad paths to cover?

**Review by Struchid — APPROVED** [review]

File: vitrina/datasets/migrations/0015_sync_dataset_relations.py
Comment (varzgalys) [reply vitrina/datasets/migrations/0015_sync_dataset_relations.py:1]: Not really, need to execute changes in to separate transactions (migrations) to avoid "pending changes" error.

File: tests/datasets/test_views.py (hunk 5 above)
Comment (varzgalys) [reply tests/datasets/test_views.py:2098]: I have tried add more test, but changes are in the middle of the flow which makes it hard to introduce new tests.
About sad paths - actually not sad paths to test. New logic is "optional", without validation or edge cases (at-least known)

File: vitrina/datasets/admin.py [hunk 6]
```diff
@@ -490,7 +491,17 @@ class GeoportalDataServiceTypeValueInline(admin.TabularInline):
     extra = 0
 
 
+class DatasetRelationAdmin(admin.ModelAdmin):
+    list_display = ("dataset", "relation", "part_of")
+    search_fields = ("dataset", "relation", "part_of")
+    list_filter = ("relation",)
+    actions = ("sync_with_mp_node",)
```
Comment (wisecrow) [inline vitrina/datasets/admin.py]: Where `sync_with_mp_node` is defined? I could not find it :(
